multiple improvements to logging

takase1121 opened the pull request:
- added style.log table
- removed contextmenu
- use ctrl+click to copy individual log entries
- use icon instead of + or - for log items in logview

I think the video demostrates the changes pretty good:

https://user-images.githubusercontent.com/20792268/146626856-e93163fc-1e5d-493c-a16d-4a212cebc755.mp4

Guldoman commented:
I really like this!
Sometimes the icons are a bit misaligned, tho.
How would the error icon look like with a different (red, orange) color?

The ctrl+click lacks a bit in discoverability, but it should be fine.

takase1121 commented:
> Sometimes the icons are a bit misaligned, tho.
~~ah crap, now I have to fix that~~ I can allocate a percentage of the View to the date, so every portion would be fixed. 

> How would the error icon look like with a different (red, orange) color?
That's what I'm thinking too, maybe we could test that out.

> The ctrl+click lacks a bit in discoverability, but it should be fine.
I want to put that somewhere, but not sure where to put it.

takase1121 commented:
I think the icon still looks offsetted. It might be the icon being in different sizes.

Guldoman commented:
You could try having a fixed width for the icons, and drawing them centered.

takase1121 commented:
> You could try having a fixed width for the icons, and drawing them centered.

That fixed it, thanks!

Jipok commented on data/core/style.lua:
```diff
@@ -66,4 +66,9 @@ style.syntax["function"] = { common.color "#93DDFA" }
 
+style.log = {
+  INFO  = { icon = "i", color = style.text },
+  ERROR = { icon = "!", color = { common.color "#ff6961" } }
```
I don't think this is a good practice.
Why not use `style.accent`?

takase1121 replied:
It's because by default style accent is just a very bright white. It doesn't look obvious nor it represents error well.

Jipok replied:
May be `style.nagbar`?
But we really do not have enough predefined colors for expressing ok/error, add/remove and the like.

Guldoman replied:
Now that we have control over the colors repo, we should add default colors for each theme for things like "info", "warning", "error", that even if we are not using directly, might be useful for plugins.

Jipok replied:
I thought about the same. Although only two colors. https://github.com/lite-xl/lite-xl/pull/749

Jipok commented:
> use ctrl+click to copy individual log entries

Can you add mouse selection just like any other text? It would be intuitive.

Guldoman commented:
> > use ctrl+click to copy individual log entries
> 
> Can you add mouse selection just like any other text? It would be intuitive.

How would this coexist with click to expand the log entry? Maybe just allowing the expand/reduce if clicked on the "main" line, and allowing selections in the entry body. This would require implementing selections in LogView tho, which I feel is a bit too much. (We could try to modularize DocView, so that it could be used for things like this)

Also, it looks like the item that gets expanded on click is not the one under the cursor during `on_mouse_pressed`, but is the last one hovered during `on_mouse_moved`.

takase1121 commented:
> Also, it looks like the item that gets expanded on click is not the one under the cursor during on_mouse_pressed, but is the last one hovered during on_mouse_moved.

Is that a problem?

> Can you add mouse selection just like any other text? It would be intuitive.
Selections are hard to implement.

Guldoman commented:
> > Also, it looks like the item that gets expanded on click is not the one under the cursor during on_mouse_pressed, but is the last one hovered during on_mouse_moved.
> 
> Is that a problem?

In some cases like when scrolling without moving the mouse, the wrong item is expanded. It's not a big problem, just a minor annoyance.

takase1121 commented:
> In some cases like when scrolling without moving the mouse, the wrong item is expanded. It's not a big problem, just a minor annoyance.

Fixed.

Also added some instruction when logview is opened, and indication when entry is copied.

Jipok commented:
> How would this coexist with click to expand the log entry?

What is the problem? There are solutions. Look at the tabs at least - you can drag and drop them, and open them.

I don't really understand why there is a separate implementation for the logs. It was possible to create a temporary document where to write all the logs. Then all his usual things would work for the user. It is not difficult to make document read-only. The only drawback compared to the current implementation of the logs would be the lack of folding. But this could prompt the implementation of this feature, which would be useful not only for logs.

Guldoman commented:
> > How would this coexist with click to expand the log entry?
> 
> What is the problem? There are solutions. Look at the tabs at least - you can drag and drop them, and open them.

Drag and drop and click are compatible actions that have been in use since forever. Clicking to toggle and dragging to select text are not very compatible.

That's why I was suggesting to use the first "header" line as a toggle, while the body of the item as a "text box"/DocView. But doing that would need a bit of restructuring.

Jipok commented:
> Drag and drop and click are compatible actions that have been in use since forever. Clicking to toggle and dragging to select text are not very compatible.

I would not say that they are not compatible. But if you don't want the element to open during selection, then just add an additional check to `on_mouse_released` so that the position matches `on_mouse_pressed`.

Jipok commented:
Missing color for tip/info type?

takase1121 commented:
> Missing color for tip/info type? 

What color you think will be suitable? to me, `style.good`?

Jipok commented:
> What color you think will be suitable?

For a regular log item? The current gray seems normal to me. By the way, I found several problems:

<details><summary>Text overlaps</summary>

![image](https://user-images.githubusercontent.com/25588359/146937585-ba5d87b8-5454-493e-9b9e-e1ac724361d9.png)

</details>
<details><summary>Large gap</summary>

![image](https://user-images.githubusercontent.com/25588359/146937855-4c87fed1-434b-442d-855f-faaf60c8709a.png)

</details>

Since you are improving the viewing of logs. Maybe add a jump to the first error when opening the logs?

Jipok commented on data/core/commands/log.lua:
```diff
@@ -0,0 +1,14 @@
+  ["log:open-as-doc"] = function()
+    local doc = core.open_doc("logs.txt")
+    core.root_view:open_doc(doc)
+    doc:insert(1, 1, core.get_log())
```
I suggest adding something like this: 
```lua
doc.new_file = false
doc:clean()
```
Allows you to close the logs and exit the editor without annoying warning about an unsaved file. At the same time, it does not interfere with actually saving as a file

takase1121 commented:
> UPD: The current command set is slightly inconsistent. We have Core: Open Log, but Log:Open As Doc and Log: Copy To Clipboard

This unfortunately can be potential backwards-compatibility issue.

adamharrison commented:
I think we're good to merge this, yes? It looks good to me.
